src/util/result.h

```diff
@@ -7,42 +7,119 @@
+//! dereferenced regardless of the function's success or failure.
+//!
+//! This class is not intended to be used by low-level functions that do not
+//! return error or warning strings. These functions should use plain
+//! std::optional or std::variant types instead.
+//!
+//! See unit tests in result_tests.cpp for example usages.
+template<typename T>
+class Result;
+
+//! Tag types for result constructors.
+struct Error{};
+struct ErrorChain{};
 
-    /* Whether the function succeeded or not */
-    bool HasRes() const { return std::holds_alternative<T>(m_variant); }
+//! Result<void> specialization. Only returns errors and warnings, no values.
+template<>
+class Result<void>
+{
+protected:
+    std::vector<bilingual_str> m_errors;
+    std::vector<bilingual_str> m_warnings;
 
-    /* In case of success, the result object */
-    const T& GetObj() const {
-        assert(HasRes());
-        return std::get<T>(m_variant);
+public:
+    //! Success case constructor.
+    Result() = default;
+
+    //! Error case constructor for a single error.
+    Result(Error, bilingual_str error)
+    {
+        m_errors.emplace_back(std::move(error));
     }
-    T ReleaseObj()
+
+    //! Error case constructor for a chained error.
+    Result(ErrorChain, bilingual_str error, Result<void>&& previous) : m_errors{std::move(previous.m_errors)}, m_warnings{std::move(previous.m_warnings)}
     {
-        assert(HasRes());
-        return std::move(std::get<T>(m_variant));
+        m_errors.emplace_back(std::move(error));
     }
 
-    /* In case of failure, the error cause */
-    const bilingual_str& GetError() const {
-        assert(!HasRes());
-        return std::get<bilingual_str>(m_variant);
+    //! Success check.
+    operator bool() const { return m_errors.empty(); }
+
+    //! Error retrieval.
+    const std::vector<bilingual_str>& GetErrors() const { return m_errors; }
+    std::tuple<const std::vector<bilingual_str>&, const std::vector<bilingual_str>&> GetErrorsAndWarnings() const { return {m_errors, m_warnings}; }
+};
+
+template<typename T>
+class Result : public Result<void>
+{
+protected:
+    T m_result;
+
+public:
+    //! Constructors that forward to the base class and pass additional arguments to m_result.
+    template<typename... Args>
+    Result(Args&&... args) : m_result{std::forward<Args>(args)...} {}
+    template<typename Str, typename...Args>
+    Result(Error, Str&& str, Args&&... args) : Result<void>{Error{}, std::forward<Str>(str)}, m_result{std::forward<Args>(args)...} {};
+    template<typename Str, typename Prev, typename...Args>
+    Result(ErrorChain, Str&& str, Prev&& prev, Args&&... args) : Result<void>{ErrorChain{}, std::forward<Str>(str), std::forward<Prev>(prev)}, m_result{std::forward<Args>(args)...} {};
+
+    //! std::optional methods, so Result<T> can be easily swapped for
+    //! std::optional<T> to add error reporting to existing code or remove it if
```

Comment:
I don't think this is true? `std::optional` must not be dereferenced, whereas `Result` can be, and would return a default constructed object?

Reply:
> I don't think this is true? `std::optional` with `nullopt` must not be dereferenced, whereas `Result` can be, and would return a default constructed object?

I can clarify comment, but this interface is superset of std::optional interface and allows dereferencing in cases when original object can't be derefenced. The object does not have to be default constructable since util::Error{} can forward any constructor arguments